reps/2022-04-21-object_store_HA.md

```diff
@@ -0,0 +1,132 @@
+- Can not recover the object returned by actor task.
+- Require task is idempotent.
+
+#### Goal
+
+1. Objects can be specified for high availability mode, other objects are the same as before.
+2. Any high availability objects should still be accessible if we encounter a single-node failure.
+
+### Should this change be within `ray` or outside?
+
+Changes are within Ray core.
+
+## Stewardship
+
+### Required Reviewers
+
+@stephanie-wang, @ericl, @scv119, @kfstorm, @raulchen
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@ericl, @raulchen, @stephanie-wang
+
+## Design and Architecture
+
+### Problem statement
+
+#### Problem 1: Object Owner Failure
+
+The owner of an object stores the metadata of the object, such as reference count and locations of the object. If the owner dies, other workers which hold the object ref cannot access the data of the object anymore because all copies of the object will be deleted from the Object Store.
+
+#### Problem 2: Object Borrower Failure
+
+In the chained object borrowing case, the owner of an object is not aware of the indirect borrowers. If all direct borrower fails, the owner will consider the object out-of-scope and GC the object. Accessing the object on indirect borrowers will fail with an `ObjectLostError`.
+
+more details: [issues 18456](https://github.com/ray-project/ray/issues/18456)
+
+#### Problem 3: Loss of All Copies
+
+Data of objects stored in the plasma store. For now, the plasma store is a thread of the raylet process, failure of the raylet process will lose data which store in plasma. Some objects which only one copy in that failed plasma store,  will be unavailable.
+
+### Proposed Design
+
+We implement
+#### Options to implement object HA with checkpoint
+
+We implement object HA based on the checkpoint, so we can walk around **Problem 3: Loss of All Copies**,
+previously discussed options: https://github.com/ray-project/enhancements/pull/10#issuecomment-1127719640
+
+We use highly available processes as global owners of checkpointed objects. Such highly available processes can be GCS or a group of named actors with `max_restarts=-1`. We reuse the existing ownership assignment RPCs to assign a checkpointed object to a global owner and encode the immutable info (an `owner_is_gcs` flag or the actor name) about the global owner into the owner address. The process to get an RPC client to the owner needs to be updated to be able to return a working RPC client to the up-to-date IP:port of the owner.
+
+Note that we don't need to restore the reference table in global owners by pulling info from the cluster because objects are already checkpointed. Checkpoint info is stored in the reference table and it will be encoded when serializing an object ref, hence checkpoint info is recoverable. If borrowers detected owner failure, they will try to reconnect to the owner and the recovered owner will recover the reference count and borrower list via these new RPC connections.
```

Comment:
Does this mean that the checkpoint location will be stored with the serialized ObjectRef? It makes sense, but if that is the idea, could you make that more explicit somewhere in the design? For example, how does the protocol for retrieving the object value change when using HA?

I'm also wondering if we can try a version where the application has to manually free the object. Then we don't need to worry about keeping track of the ref count at all. Would this meet your use cases?

Reply:
yes, we will store checkpoint_uri into ObjectRef, protocol is here: https://github.com/ray-project/enhancements/pull/10#discussion_r948825477. I will add to the REP later.

yes, In our prototype test, we do not recover the reference count after the owner restart, because the HA Object doesn't have a primary copy, and users need to release the checkpoint by themself.

Reply:
In the current prototype, refs caused by borrowers are still recorded. We didn't simplify this. But if the global owner is recovered from failure, the reference count will become zero and direct borrowers will not re-report borrow info to the new owner process. In this case, the owner will not GC the object because it knows that the object is HA.

We did think about bypassing reference recording for HA objects. But for the simplicity of the prototype, we didn't implement this.